File: problems/python/arrays/two_sum.py
"""
Two Sum
Given an array of integers nums and an integer target,
return indices of the two numbers such that they add up to target.
"""

from typing import List


def two_sum(nums: List[int], target: int) -> List[int]:
    """
    Find two numbers that add up to target.

    Args:
        nums: List of integers
        target: Target sum

    Returns:
        List of two indices
    """
    num_map = {}

    for i, num in enumerate(nums):
        complement = target - num
        if complement in num_map:
            return [num_map[complement], i]
        num_map[num] = i

    return []


if __name__ == "__main__":
    # Example usage
    nums = [2, 7, 11, 15]
    target = 9
    result = two_sum(nums, target)
    print(f"Two Sum Example:")
    print(f"Input: nums = {nums}, target = {target}")
    print(f"Output: {result}")
    print(
        f"Explanation: nums[{result[0]}] + nums[{result[1]}] = {nums[result[0]]} + {nums[result[1]]} = {target}"
    )

File: problems/c/arrays/two_sum.c
/*
 * Two Sum
 * Given an array of integers nums and an integer target,
 * return indices of the two numbers such that they add up to target.
 */

#include <stdio.h>
#include <stdlib.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* twoSum(int* nums, int numsSize, int target, int* returnSize) {
    int* result = (int*)malloc(2 * sizeof(int));
    *returnSize = 2;
    
    for (int i = 0; i < numsSize; i++) {
        for (int j = i + 1; j < numsSize; j++) {
            if (nums[i] + nums[j] == target) {
                result[0] = i;
                result[1] = j;
                return result;
            }
        }
    }
    
    *returnSize = 0;
    return result;
}

int main() {
    int nums[] = {2, 7, 11, 15};
    int target = 9;
    int returnSize;
    
    int* result = twoSum(nums, 4, target, &returnSize);
    
    if (returnSize == 2) {
        printf("Two Sum Example:\n");
        printf("Input: nums = [2, 7, 11, 15], target = 9\n");
        printf("Output: [%d, %d]\n", result[0], result[1]);
        printf("Explanation: nums[%d] + nums[%d] = %d + %d = %d\n",
               result[0], result[1], nums[result[0]], nums[result[1]], target);
    }
    
    free(result);
    return 0;
}
